lib/integrations/awsoidc/opentunnelec2.go

```diff
@@ -0,0 +1,402 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	go func() {
+		defer ec2Conn.Close()
+
+		logger := log.WithField("aws-oidc", "opentunnel").
+			WithField("instance_id", req.InstanceID).
+			WithField("eice", *eice.InstanceConnectEndpointId)
+
+		local, err := req.Listener.Accept()
+		if err != nil {
+			logger.WithError(err).Warn("failed to accept connection")
+			return
+		}
+		defer local.Close()
+
+		// Only one client is expected, no need to keep listener around.
+		req.Listener.Close()
+
+		g, _ := errgroup.WithContext(ctx)
+		g.Go(func() error {
+			_, err := io.Copy(local, ec2Conn)
+			return trace.Wrap(err, "dst:local src:ec2")
+		})
+		g.Go(func() error {
+			_, err := io.Copy(ec2Conn, local)
+			return trace.Wrap(err, "dst:ec2 src:local")
+		})
+
+		if err := g.Wait(); err != nil {
+			logger.WithError(err).Warn("failed to copy data")
+		}
+	}()
+
+	return ret, nil
+}
+
+// fetchEC2InstanceConnectEndpoint returns an EC2InstanceConnectEndpoint for the given VPC and whose state is ready to use ("create-complete").
+func fetchEC2InstanceConnectEndpoint(ctx context.Context, clt OpenTunnelEC2Client, req OpenTunnelEC2Request) (*ec2types.Ec2InstanceConnectEndpoint, error) {
+	describe, err := clt.DescribeInstanceConnectEndpoints(ctx, &ec2.DescribeInstanceConnectEndpointsInput{
+		Filters: []ec2types.Filter{
+			{
+				Name:   &filterEC2InstanceConnectEndpointVPCIDKey,
+				Values: []string{req.VPCID},
+			},
+			{
+				Name:   &filterEC2InstanceConnectEndpointStateKey,
+				Values: []string{string(ec2types.Ec2InstanceConnectEndpointStateCreateComplete)},
+			},
+		},
+	})
+	if err != nil {
+		return nil, trace.BadParameter("failed to list EC2 Instance Connect Endpoint for VPC %q (region=%q): %v", req.VPCID, req.Region, err)
+	}
+
+	if len(describe.InstanceConnectEndpoints) == 0 {
+		return nil, trace.BadParameter("no EC2 Instance Connect Endpoint for VPC %q (region=%q), please create one", req.VPCID, req.Region)
+	}
+
+	return &describe.InstanceConnectEndpoints[0], nil
+}
+
+// dialEC2InstanceUsingEICERequest is a request to dial into an EC2 Instance Connect Endpoint.
+type dialEC2InstanceUsingEICERequest struct {
+	credsProvider    aws.CredentialsProvider
+	awsRegion        string
+	endpointId       string
+	customCA         *x509.Certificate
+	endpointHost     string
+	privateIPAddress string
+	remotePort       string
+}
+
+// dialEC2InstanceUsingEICE dials into an EC2 instance port using an EC2 Instance Connect Endpoint.
+// Returns a net.Conn that transparently proxies the connection to the EC2 instance.
+func dialEC2InstanceUsingEICE(ctx context.Context, req dialEC2InstanceUsingEICERequest) (net.Conn, error) {
+	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/connect-using-eice.html
+	q := url.Values{}
+	q.Set("instanceConnectEndpointId", req.endpointId)
+	q.Set("maxTunnelDuration", "3600") // 1 hour (max allowed)
+	q.Set("privateIpAddress", req.privateIPAddress)
+	q.Set("remotePort", req.remotePort)
```

Comment:
Comments should clarify that this is inferred from the aws cli source rather than being an officially documented API.  I imagine that it's fairly stable since since changing this would break old CLI clients, but still best to make sure the risk is well-documented.

Reply:
Updated the comment and linked to the awscli implementation